### PR TITLE
feat(backend): report, hide and clean up spray walls; rollout telemetry and docs (SW-17)

### LINKED ISSUE #5450: SW-17: Ops, safety, rollout, docs
Epic #5346 (SW-17). Safety rails, telemetry, the flag rollout and the docs that let the next person maintain this.

## Why

User-uploaded photos are a new class of content; walls consume storage that has to be reclaimed; the feature ships behind a flag and needs the events that tell us whether detection and resets actually work in the wild. PostHog is past the 1M-event tier, so instrument outcomes, not gestures.

## Scope

- `reportSprayWall(wallUuid, reason)` writing a small flag table with an admin `hidden` toggle (the `climb_proposals` vote machinery in `docs/climb-moderation.md` is climb-shaped and a poor fit); hidden walls disappear from gym lists, the web view and share links but stay visible to the owner with a notice.
- Scheduler job (`packages/scheduler`, `docs/scheduler.md`): 30 days after a wall's soft delete, remove `spray-walls/<wallUuid>/` from both buckets and hard-delete the `spray_walls` side rows (catalogue rows and climbs persist by design). Registry test pinned.
- Caps from `spray-config.ts` surfaced as copy in the app (walls per user, holds per wall, versions per wall) and enforced server-side (SW-05).
- Telemetry in `packages/shared/analytics/src/events.ts` (builder style of `board-render-events.ts`; no free text, no coordinates): `Spray Wall Photo Picked` {source, width, height, bytes}, `Spray Wall Upload Finished` {ms, ok, error_reason}, `Spray Wall Detection Finished` {hold_count, model_version, ms, tiles}, `Spray Holds Reviewed` {detected, added, moved, deleted, resized, accepted_all, ms}, `Board Created` (existing) with `boardType: 'spray'`, `Spray Wall Reset Previewed` {kept, removed, added, climbs_affected}, `Spray Wall Reset Applied`, `Climb Remixed From Broken` {lost_hold_count}. `Board Render Failed` already carries `board_name`. Wire the calls in SW-09 / SW-13 if they landed without them.
- Flag rollout plan: `spray-walls` on for testers → 10 % → everyone, with the PostHog insight that gates each step (detection F1 proxy = `Spray Holds Reviewed` corrections / detected ≤ 0.15; reset satisfaction = commits / previews).
- Docs: `docs/spray-walls.md` finished (identity mapping, tables, canonical space, detection, resets, privacy, ops), `docs/entity-structure.md` gains the wall entity, CLAUDE.md docs index line, `docs/feature-flags.md` mobile flag row.

## Out of scope

Automated image moderation (a manual report queue is enough at this volume), storage quotas per gym.

## Acceptance

- Reporting a public wall hides it after an admin toggles `hidden`; the owner still sees it with the notice.
- The cleanup job removes objects and rows for a wall deleted 31 days ago and leaves a wall deleted yesterday alone (test with a frozen clock).
- Every event above appears in PostHog from a dev-client run with the documented properties and nothing else.
- Docs reviewed by a second agent for accuracy against the merged code.

## Where

`packages/backend/src/graphql/resolvers/board/spray-walls.ts`, `packages/db/src/schema/app/spray-walls.ts` (flag table), `packages/scheduler/src/jobs/`, `packages/shared/analytics/src/events.ts`, `docs/spray-walls.md`, `docs/entity-structure.md`, `docs/feature-flags.md`, `CLAUDE.md`.

## Blocked by / Blocks

- Blocked by: SW-13 (#5446).
- Blocks: none.

## Size

M.

### PR DESCRIPTION
**Pull request by marcodejongh**
## Summary

The ops half of spray walls: what happens when a photograph of somebody's home turns out to be a problem, and what happens to it after the wall is deleted.

**Reporting and hiding.** `reportSprayWall` writes one `spray_wall_reports` row per (wall, reporter) — a second report from the same climber answers `ALREADY_REPORTED` and writes nothing — and `setSprayWallHidden` is the admin switch. Deliberately not the `climb_proposals` vote machinery: a proposal is a change to a climb that several climbers who have been on it are the right judges of, while "should we be serving this photograph" has one right answer and the only people who could vote are the ones who can already see the wall. So it is a queue an admin reads and one flag.

A hidden wall reads **exactly like a private one for everybody but its owner**, who keeps it with a notice. Hiding outranks both `is_public` and the unlisted share-link capability — it has to take a wall off the internet, and a link that still worked would not be that. The rule lands in four implementations that do not share a query builder: the two resolver helpers, `viewerCanWriteSprayClimbs`, and the three SQL predicates in `spray-visibility.ts` that gate a spray climb's ~15 reads. A test asserts the by-uuid read, the by-layout read and the climb read separately, plus that the owner still has all three.

**Retention.** Thirty days after an owner deletes a wall, `purge-spray-wall-photos` (07:00 UTC) calls a cron-authenticated mutation that deletes `spray-walls/<wallUuid>/` in the private bucket and in `media` (where SW-14's public promotion copies a published photo), then clears `photo_key`. Objects first, row second: a crash between them leaves a version with no photo, which is what the next run would have produced anyway, where the other order orphans an object nothing names. No wall, hold or climb row is ever deleted — other people's ticks point at them.

**Caps, said out loud.** Ten walls, 1,500 holds, fifty resets are all reachable by ordinary use, so each is stated with its number — a hint on the create step before it bites, the editor's save refusal, the reset's upload failure — read from `spray-config.ts` rather than typed into a catalog string, and matched on `extensions.code` rather than the server's untranslated English.

**Telemetry.** The seven events already existed as inline object literals at six call sites. They now go through typed builders in the `board-render-events.ts` style, so "outcomes only, and nothing that identifies a wall" is a property of the code rather than of whoever writes the next call site. A test reads every payload back field by field.

**Rollout.** `spray-walls` goes testers → 10 % → everyone, each step gated on a ratio that lives as a function in `spray-wall-events.ts` rather than as prose in a dashboard description, so the doc and the code cannot drift.

**Reporting ships as API only.** `reportSprayWall`, `setSprayWallHidden` and `sprayWallReports` all work; the only client wiring is the owner's hidden-wall banner. A report row belongs in `BoardDetailSheet`, which SW-11's stack is rewriting, so putting one there from this stack would have been a guaranteed conflict for no gain — the admin path needs the mutation either way. Follow-up: #5501 (SW-17b, in-app Report button + admin review route). The test plan above is what a tester can actually reach today.

Note for the migration bot: SW-14 (#5496) also numbers its migration 0229, on the other stack. Whichever lands first, the other gets `vp run db:renumber`.

Stacked on #5492 (SW-13) → #5484 → #5481 → #5479, merging #5480; merge those first, then this PR retargets to main.

## Release Notes

The app now tells you a wall's limits up front instead of after you hit them: ten walls each, 1,500 holds on a wall, fifty resets. If a wall of yours ever gets taken out of everyone else's view, you'll see why on it — it stays yours, your climbs stay on it, and nothing is deleted. Delete a wall yourself and its photos come off our disks a month later.

## Test plan

1. Start a new wall → the first step names your wall limit.
2. Add holds past 1,500 → the save names the number.
3. Reset a wall on its fiftieth reset → the upload names the limit.
4. Open a wall an admin hid → a notice on it explains why.
5. That hidden wall still opens, resets and shows your climbs.

## Risk

Risk: 4/5 — new moderation flag, a migration, and a scheduled deletion job

Closes #5450

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01Q7BjHrRALJ1gjtiRWm5rLV

### CODE REVIEW

File: packages/backend/src/graphql/resolvers/board/spray-wall-moderation.ts [hunk 1]
```diff
@@ -0,0 +1,373 @@
...
+      await tx
+        .update(dbSchema.sprayWalls)
+        .set({
+          hiddenAt: validated.hidden ? (loaded.wall.hiddenAt ?? now) : null,
+          hiddenBy: validated.hidden ? (loaded.wall.hiddenBy ?? ctx.userId!) : null,
+          updatedAt: now,
+        })
+        .where(eq(dbSchema.sprayWalls.id, loaded.wall.id));
+
+      if (validated.hidden) await purgeSprayWallFeedItems(tx, loaded.wall.layoutId);
```
[the remark this answers is not in the document]
Comment (marcodejongh) [reply packages/backend/src/graphql/resolvers/board/spray-wall-moderation.ts]: Fixed. Hiding purged the feed rows that already existed and nothing stopped the next one — the wall is still `is_public`, so the next climb or tick would announce itself and put the wall straight back into the feed it was just taken out of.

Both write-time gates now read `hidden_at`: `publishesFeedEvents` in `spray-authoring.ts`, and — the one that actually decides — `sprayWallMayAnnounceUnderLock`, under the wall lock, for exactly the race that gate already existed for (`setSprayWallHidden`'s purge cannot catch a row that has not been written yet).

Test: `announces nothing for a climb set on it after it was hidden` asserts a `climb.created` before hiding and none after. Mutation-checked — dropping `&& row.hiddenAt == null` from the under-lock read reds it.

File: packages/backend/src/graphql/resolvers/board/spray-wall-moderation.ts [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+
+  // `lt`, not `lte`: a wall deleted exactly at the cutoff is inside the window by
+  // a hair, and a retention window should err towards keeping a photo one more
+  // run rather than deleting it one run early.
+  const candidates = await db
+    .select({ id: dbSchema.sprayWalls.id, boardUuid: dbSchema.sprayWalls.boardUuid })
+    .from(dbSchema.sprayWalls)
+    .where(and(isNotNull(dbSchema.sprayWalls.deletedAt), lt(dbSchema.sprayWalls.deletedAt, cutoff)))
+    .orderBy(asc(dbSchema.sprayWalls.deletedAt))
+    .limit(batchSize);
```
[the remark this answers is not in the document]
Comment (marcodejongh) [reply packages/backend/src/graphql/resolvers/board/spray-wall-moderation.ts]: Fixed, and it was worse than an inefficiency: a purged wall's row is never deleted, so it stays past the cutoff forever. With the filter after `LIMIT 200`, every run past the two-hundredth purge filled its batch with no-ops and no wall deleted afterwards was ever reached again — a permanent `wallsPurged: 0` that reads exactly like "nothing to do".

"Already purged" is now explicit state (`spray_walls.photos_purged_at`, migration 0229) inside the candidate query, with `spray_walls_deleted_at_idx` partial on `deleted_at IS NOT NULL AND photos_purged_at IS NULL`. It is stamped only after the objects are gone, so a failed sweep leaves the wall a candidate.

Test: `reaches a fresh deletion past a full batch of already-purged walls`. Mutation-checked — moving the predicate out of the WHERE reds it.

File: packages/backend/src/graphql/resolvers/board/spray-wall-moderation.ts [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+  // deleted.
+  const wallIds = candidates.map((candidate) => candidate.id);
+  const remaining = await db
+    .select({ wallId: dbSchema.sprayWallVersions.wallId })
+    .from(dbSchema.sprayWallVersions)
+    .where(and(inArray(dbSchema.sprayWallVersions.wallId, wallIds), isNotNull(dbSchema.sprayWallVersions.photoKey)))
+    .groupBy(dbSchema.sprayWallVersions.wallId);
+  const wallIdsWithPhotos = new Set(remaining.map((row) => Number(row.wallId)));
+
+  const work = candidates.filter((candidate) => wallIdsWithPhotos.has(Number(candidate.id)));
```
[the remark this answers is not in the document]
Comment (marcodejongh) [reply packages/backend/src/graphql/resolvers/board/spray-wall-moderation.ts]: Fixed. This is why the marker had to stop being "has a version with a photo key": a photo uploaded into the wizard is unreferenced until `createSprayWallVersion` adopts it, and if the climber backs out it never is — so nothing in the database names the object and a version-driven sweep left it in the bucket forever.

The sweep now lists the PREFIX (`spray-walls/<wallUuid>/`, both buckets) and the rows are merely what it clears afterwards. `photos_purged_at` is what keeps that from re-listing an empty prefix on every future run.

Test: `deletes an abandoned upload that no version ever adopted` — a wall with no version at all, one stray key, deleted 31 days ago; the run deletes exactly that key.

File: packages/backend/src/graphql/resolvers/board/spray-wall-moderation.ts [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+ *
+ * The uuid rule, not the by-layout one: a report is always made from a wall
+ * somebody is looking at, and an unlisted wall someone was sent a link to is
+ * exactly the case worth reporting. A wall a viewer cannot see answers "not
+ * found", the same as a uuid that is not a wall at all — a report must never be
+ * an oracle for which uuids exist.
+ */
+function viewerCanReport(wall: WallRow, board: BoardRow, userId: string): boolean {
+  if (wall.hiddenAt != null) return board.ownerId === userId;
+  return board.isPublic || board.isUnlisted || board.ownerId === userId;
```
[the remark this answers is not in the document]
Comment (marcodejongh) [reply packages/backend/src/graphql/resolvers/board/spray-wall-moderation.ts]: Fixed. `viewerCanReport` now delegates to `viewerCanSeeSprayWall` instead of restating the rule — restating it is how the gym-member path got dropped, and the person most likely to notice something wrong with a gym's wall is somebody who climbs there.

Test: `lets a gym member report the gym’s PRIVATE wall`, with a gym, a member row and the wall attached.

File: packages/backend/src/graphql/resolvers/board/spray-wall-moderation.ts [hunk 5]
```diff
@@ -0,0 +1,373 @@
...
+        createdAt: dbSchema.sprayWallReports.createdAt,
+        wallUuid: dbSchema.sprayWalls.boardUuid,
+        layoutId: dbSchema.sprayWalls.layoutId,
+        hiddenAt: dbSchema.sprayWalls.hiddenAt,
+      })
+      .from(dbSchema.sprayWallReports)
+      .innerJoin(dbSchema.sprayWalls, eq(dbSchema.sprayWalls.id, dbSchema.sprayWallReports.wallId))
+      .where(and(...conditions))
+      .orderBy(sql`${dbSchema.sprayWallReports.createdAt} DESC`)
+      .limit(200);
```
[the remark this answers is not in the document]
Comment (marcodejongh) [reply packages/backend/src/graphql/resolvers/board/spray-wall-moderation.ts:424]: Fixed. `sprayWallReports` now carries `isNull(sprayWalls.deletedAt)`. Excluded rather than marked reviewed on delete: the row is the record of why, and "reviewed" should mean an admin looked.

Test: `drops a pending report once the owner deletes the wall` — one pending report, then `deleteSprayWall`, then the queue is empty.

**Comment by marcodejongh** [comment]
## Review items addressed

Every distinct item across the seven claude[bot] reviews, with where it landed. Code was pushed: head is now `3473acd7f`.

### Fixed in `3473acd7f` (this round)

- [x] **Upload retry could loop on a failed version check** (`SprayWallWizardScreen.tsx`). Correct, and the Codex round did *not* cover it — that round added the reconcile query, this one covers the query itself failing. Extracted to `planUploadRetry` (`resume-draft.ts`) with three answers: `adopt` a draft a lost response already created, `upload` when the wall genuinely has nothing, `blocked` when the wall could not be *read*. `blocked` surfaces `sprayWizard.upload.checkFailed` instead of re-uploading into a guaranteed refusal. Tested (4 cases) and mutation-tested — flipping `blocked` back to `upload` fails one.
- [x] **`useSprayWallBuilder` return not memoised** (CLAUDE.md mobile checklist). Wrapped in `useMemo`. Two tests: stable identity across a re-render, new identity when a held value changes.
- [x] **`BACK_TARGET.detect` hazard**. Entry removed. Backing into `photo` with a draft open would hit `runUpload`'s draft guard and sit there doing nothing — worse than no way back. `upload` keeps its way back, because a draft cannot exist on that step. Tested.
- [x] **`findResumableWall` picks by server order**. The comment was wrong about recency: `mySprayWalls` orders `spray_walls.created_at` DESC (`resolvers/board/spray-walls.ts`), so the first unfinished wall *is* the most recent. Comment corrected to cite the ordering; no client-side re-sort, because the wall payload carries no timestamp to sort by.
- [x] **Missing test for the null-history retry path**. That is the `planUploadRetry` suite above — a pure helper rather than a screen test, so the branch is pinned rather than merely exercised.

### Already fixed at `f01cf4751` or earlier

- [x] **i18n `{{value}}` vs `{{count}}`** — `detectionSummary` passes `{ value }`; `review-summary-copy.test.ts` renders the real catalog in all four locales and one case pins that the wrong name leaves the placeholder, so the assertion is not vacuous.
- [x] **`SprayCornerMarker` handles frozen after Clear** — effect keyed on the seed's values writes the shared values back.
- [x] **`parseVersion` dead `length === 0` branch** — removed.
- [x] **Division by zero / NaN in the preview height** — `previewHeight()` guards both dimensions and falls back to 4:3.
- [x] **`docs/spray-walls.md` said the UI did not exist** — new "Adding a wall" section (step table plus the three non-obvious rules), and the three stale lines corrected.
- [x] **Public wall discoverable before its first publish** — created private always; visibility applied by `updateSprayWall` after the `PUBLISHED` latch.
- [x] **Orphan wall from an abandoned run** — the flow opens by offering an unfinished wall back, or discarding it.

### Answered rather than changed

- **`BoardVisibilityFields` mounts `useDeviceLocation` unconditionally.** Correct, and left as is: the hook requests no permission on mount, and `location.coords` stays null until the climber taps "Use my location" — the same contract `BoardForm` relied on behind its disclosure. Gating it on a disclosure the wizard does not have would mean a second code path for no behavioural difference.
- **A wall row can exist before any version** (raised as a note, not a bug). True by construction — the photo handler authorises against the wall — and that is exactly what the resume check now exists for.
- **A retry after a *non-committed* `createSprayWallVersion` re-uploads the photo.** Unavoidable: nothing distinguishes "did not commit" from "committed, response lost" except asking, and when the ask succeeds `planUploadRetry` adopts. The worst case is one unreferenced object, which the SW-17 cleanup job sweeps — it is never a row anybody can see.
- **No render smoke test for `SprayWallWizardScreen`.** Agreed and deferred to its own follow-up, at the coordinator's call: it needs the boards-index-sized mock set and is better landed on its own than bolted onto a fix commit.

### Gates at `3473acd7f`

`vp run typecheck:mobile` clean · `spray` suites 20 files / **305 tests** · `boards` 29 files / **407 tests** · `vp check` 0 errors, 370 warnings (baseline).
